lib/trino-parquet/src/test/java/io/trino/parquet/reader/TestParquetReaderMemoryUsage.java

```diff
@@ -0,0 +1,208 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.parquet.reader;
+
+import com.google.common.collect.ImmutableList;
+import io.airlift.units.DataSize;
+import io.trino.memory.context.AggregatedMemoryContext;
+import io.trino.parquet.Field;
+import io.trino.parquet.ParquetDataSource;
+import io.trino.parquet.ParquetReaderOptions;
+import io.trino.parquet.writer.ParquetSchemaConverter;
+import io.trino.parquet.writer.ParquetWriter;
+import io.trino.parquet.writer.ParquetWriterOptions;
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.LazyBlock;
+import io.trino.spi.type.Type;
+import org.apache.parquet.hadoop.metadata.BlockMetaData;
+import org.apache.parquet.hadoop.metadata.CompressionCodecName;
+import org.apache.parquet.hadoop.metadata.ParquetMetadata;
+import org.apache.parquet.io.MessageColumnIO;
+import org.joda.time.DateTimeZone;
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.memory.context.AggregatedMemoryContext.newSimpleAggregatedMemoryContext;
+import static io.trino.parquet.ParquetTypeUtils.constructField;
+import static io.trino.parquet.ParquetTypeUtils.getColumnIO;
+import static io.trino.parquet.ParquetTypeUtils.lookupColumnByName;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.TypeUtils.writeNativeValue;
+import static java.util.Collections.nCopies;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.joda.time.DateTimeZone.UTC;
+
+public class TestParquetReaderMemoryUsage
+{
+    @Test
+    public void testColumnReaderMemoryUsage()
+            throws IOException
+    {
+        File outputFile = Files.createTempFile("memory_usage_test", "").toFile();
+        outputFile.deleteOnExit();
+        // Write a file with 100 rows per row-group
+        List<String> columnNames = ImmutableList.of("columnA", "columnB");
+        List<Type> types = ImmutableList.of(INTEGER, BIGINT);
+        writeParquetFile(outputFile, types, columnNames);
+
+        FileParquetDataSource dataSource = new FileParquetDataSource(outputFile, new ParquetReaderOptions());
```

Comment:
What is the benefit of using `FileParquetDataSource ` over `TestingParquetDataSource`?

Reply:
It's more straightforward to have the parquet writer write to file and read from it, even though it is possible to use a ByteArrayOutputStream.
Though the amount of data is small here, we will want to have the ability to read from file for other tests that might work with more data in the future.

Reply:
you are adding a new class (`FileParquetDataSource`) + requirement to have disk space available + the file creation code pollutes the test case and gives the impression the file is needed for some reason.
I would drop `FileParquetDataSource` and reintroduce it once it's needed.